piker/data/_buffer.py

```diff
@@ -54,62 +60,92 @@ async def increment_ohlc_buffer(
     the underlying buffers will actually be incremented.
     """
 
-    # wait for brokerd to signal we should start sampling
-    await shm_incrementing(shm_token['shm_name']).wait()
+    # # wait for brokerd to signal we should start sampling
+    # await shm_incrementing(shm_token['shm_name']).wait()
 
     # TODO: right now we'll spin printing bars if the last time stamp is
     # before a large period of no market activity.  Likely the best way
     # to solve this is to make this task aware of the instrument's
     # tradable hours?
 
+    global _incrementers
+
     # adjust delay to compensate for trio processing time
     ad = min(_shms.keys()) - 0.001
 
     total_s = 0  # total seconds counted
     lowest = min(_shms.keys())
     ad = lowest - 0.001
 
-    while True:
-        # TODO: do we want to support dynamically
-        # adding a "lower" lowest increment period?
-        await trio.sleep(ad)
-        total_s += lowest
-
-        # increment all subscribed shm arrays
-        # TODO: this in ``numba``
-        for delay_s, shms in _shms.items():
-            if total_s % delay_s != 0:
-                continue
-
-            # TODO: numa this!
-            for shm in shms:
-                # TODO: in theory we could make this faster by copying the
-                # "last" readable value into the underlying larger buffer's
-                # next value and then incrementing the counter instead of
-                # using ``.push()``?
-
-                # append new entry to buffer thus "incrementing" the bar
-                array = shm.array
-                last = array[-1:][shm._write_fields].copy()
-                # (index, t, close) = last[0][['index', 'time', 'close']]
-                (t, close) = last[0][['time', 'close']]
-
-                # this copies non-std fields (eg. vwap) from the last datum
-                last[
-                    ['time', 'volume', 'open', 'high', 'low', 'close']
-                ][0] = (t + delay_s, 0, close, close, close, close)
-
-                # write to the buffer
-                shm.push(last)
-
-        # broadcast the buffer index step
-        yield {'index': shm._last.value}
-
-
-def subscribe_ohlc_for_increment(
-    shm: ShmArray,
-    delay: int,
+    with trio.CancelScope() as cs:
+
+        # register this time period step as active
+        _incrementers[delay_s] = cs
+        task_status.started(cs)
+
+        while True:
+            # TODO: do we want to support dynamically
+            # adding a "lower" lowest increment period?
+            await trio.sleep(ad)
+            total_s += lowest
+
+            # increment all subscribed shm arrays
+            # TODO: this in ``numba``
+            for delay_s, shms in _shms.items():
+                if total_s % delay_s != 0:
+                    continue
+
+                # TODO: numa this!
```

Comment:
Ahh right, that was the last thing we should do. Stick this in a new `_sampling.py` module instead of the current `data._buffer` which is a lame name.